README.md
# mpesa

A dart wrapper around mpesa daraja api.

Ready Methods/APIs

- [x] LIPA NA MPESA
- [ ] C2BSIMULATE
- [ ] B2B
- [ ] C2B
- [ ] B2C
- [ ] TRANSACTION STATUS
- [ ] ACCOUNT BALANCE
- [ ] REVERSAL

## Requisites

You Will need a few things from Safaricom before development.

1. Consumer Key
2. Consumer Secret
3. Test Credentials for Development/Sanbox environment

- Login or Register as a Safaricom developer [here](https://developer.safaricom.co.ke/) if you haven't.
- Add a new App [here](https://developer.safaricom.co.ke/MyApps)
- You will be issued with a Consumer Key and Consumer Secret. You will use these to initiate an Mpesa Instance.
- Obtain Test Credentials [here](https://developer.safaricom.co.ke/APIs/MpesaExpressSimulate).
  - The Test Credentials Obtained Are only valid in Sandbox/Development environment. Take note of them.
  - To run in Production Environment you will need real Credentials.
    - To go Live and be issued with real credentials,please refer to [this guide](https://developer.safaricom.co.ke/Documentation)

## Getting Started

Add dependency in pubspec.yaml

```yaml
dependencies:
  mpesa: [ADD_LATEST_VERSION_HERE]
```

Import in your Flutter app or plain dart app.

```dart
import 'package:mpesa/mpesa.dart';

class MYClass {

    Mpesa mpesa = Mpesa(
        clientKey: "YOUR_CONSUMER_KEY_HERE",
        clientSecret: "YOUR_CONSUMER_SECRET_HERE",
        passKey: "YOUR_LNM_PASS_KEY_HERE",
        environment: "sandbox",
    );

}
```

Environment should be either `sandbox` or `production`

## Methods and Api Calls

### Lipa Na Mpesa Online (LNMO)

Lipa na M-Pesa Online Payment API is used to initiate a M-Pesa transaction on behalf of a customer using STK Push. This is the same technique mySafaricom App uses whenever the app is used to make payments.

```dart
import 'package:mpesa/mpesa.dart';

class MYClass {

    myMethod(){
        mpesa
            .lipaNaMpesa(
                phoneNumber: "",
                amount: 1,
                businessShortCode: "",
                callbackUrl: "",
            )
            .then((result) {})
            .catchError((error) {});
    }
}
```

1. businessShortCode - The organization shortcode used to receive the transaction.
2. amount - The amount to be transacted.
3. phoneNumber - The MSISDN sending the funds.
4. callbackUrl - The url to where responses from M-Pesa will be sent to.

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/newtonmunene99/mpesa-dart/issues

## Recording

![Recording](https://github.com/newtonmunene99/mpesa-dart/blob/master/recording.gif)

## Demo App

[Demo App](https://github.com/newtonmunene99/flutter_mpesa_demo)

## Contributing

1. Fork the project then clone the forked project
2. Create your feature branch: `git checkout -b my-new-feature`
3. Make your changes and add name to Contributors list below and in authors in pubspec.yaml
4. Commit your changes: `git commit -m 'Add some feature'`
5. Push to the branch: `git push origin my-new-feature`
6. Submit a pull request.

## Credits

| Contributors                                       |
| -------------------------------------------------- |
| [Newton Munene](https://github.com/newtonmunene99) |

----------------------------

For help getting started with Flutter, view their
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
# mpesa

Mpesa package changelog

## [2.0.2] - 18/01/2022

* Update README.md

## [2.0.1] - 18/01/2022

* Fix issue [#4](https://github.com/newtonmunene99/mpesa-dart/issues/4)

## [2.0.0] - 24/05/2021

* `clientKey`,`clientSecret` and `environment` are now public.
* `securityCredential` is no longer required for Lipa Na Mpesa Online

### Breaking Changes

* Migrate to null-safety
* Requires dart version `>=2.12.0 <3.0.0`
* Calling `lipaNaMpesa` method now returns `Future<MpesaResponse>` instead of `Future<Map<dynamic,dynamic>>`
  
## [1.0.0] - 29/8/2019

* Stable release
* Shipped Lipa Na Mpesa method
* Improved docs

## [0.0.4-dev.1] - 29/8/2019

* Rewrite to support plain dart without flutter

## [0.0.3] - 14/8/2019

* Some tweaking.

## [0.0.2] - 14/8/2019

* Added barebones Lipa na Mpesa method

## [0.0.1] - 14/8/2019

* Init to win it


test/mpesa_test.dart
import 'package:mpesa/mpesa.dart';
import 'package:test/test.dart';

void main() {
  group('', () {
    late Mpesa mpesa;

    setUp(() {
      mpesa = Mpesa(
        clientKey: "",
        clientSecret: "",
        environment: "sandbox",
        passKey: "",
      );
    });

    test('is instance of Mpesa', () {
      expect(mpesa, isA<Mpesa>());
    });

    test('client key is not null', () {
      expect(mpesa.clientKey, isNotNull);
    });

    test('client secret is not null', () {
      expect(mpesa.clientSecret, isNotNull);
    });

    test('environment is not null', () {
      expect(mpesa.environment, isNotNull);
    });

//was throwing this error before
//type 'String' is not a subtype of type 'int' in type cast

    test(
        "testing MpesaResponse fromMap  constructor method response code casting",
        () {
      final Map<String, dynamic> demoResponse = {
        "MerchantRequestID": "7623-624922-1",
        "CheckoutRequestID": "ws_CO_110720211339362008",
        "ResponseCode": "0",
        "ResponseDescription": "Success. Request accepted for processing,",
        "CustomerMessage": "Success. Request accepted for processin"
      };
      final MpesaResponse mpesaResponse = MpesaResponse.fromMap(demoResponse);

      expect(mpesaResponse.ResponseCode, 0);
    });
  });
}


demo/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:mpesa_test/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


demo/lib/main.dart
import 'package:flutter/material.dart';
import 'package:mpesa/mpesa.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  Mpesa _mpesa = Mpesa(
    clientKey: "qPB2VrkOdN0mHzwuKbfBdxghTB8BxDB0",
    clientSecret: "JcmP4DBFZc3CjXo0",
    passKey: "bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919",
    initiatorPassword: "Safaricom007@",
    environment: "sandbox",
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        child: Center(
          child: RaisedButton(
            child: Text("PAY"),
            onPressed: () {
              _mpesa
                  .lipaNaMpesa(
                    phoneNumber: "254792274634",
                    amount: 1,
                    businessShortCode: "174379",
                    callbackUrl:
                        "https://darajaendpoint-git-master.newtonmunene99.now.sh/callback",
                  )
                  .then((result) {})
                  .catchError((error) {});
            },
          ),
        ),
      ),
    );
  }
}


lib/utils/routes.dart
/// Holds all routes required
class Routes {
  /// Production environment base url
  static const String production = "https://api.safaricom.co.ke";

  /// Sandbox environment base url
  static const String sandbox = "https://sandbox.safaricom.co.ke";

  /// OAuth endpoint
  static const String oauth =
      "/oauth/v1/generate?grant_type=client_credentials";

  /// Business to Consumer(B2C) endpoint
  static const String b2c = "/mpesa/b2c/v1/paymentrequest";

  /// Business to Business(B2B) endpoint
  static const String b2b = "/mpesa/b2b/v1/paymentrequest";

  /// Customer to Business(C2B) register endpoint
  static const String c2bregister = "/mpesa/c2b/v1/registerurl";

  /// Customer to Business(C2B) simulate endpoint
  static const String c2bsimulate = "/mpesa/c2b/v1/simulate";

  /// Account balance endpoint
  static const String accountbalance = "/mpesa/accountbalance/v1/query";

  /// Transaction status endpoint
  static const String transactionstatus = "/mpesa/transactionstatus/v1/query";

  /// Reversal endpoint
  static const String reversal = "/mpesa/reversal/v1/request";

  /// Stk push endpoint
  static const String stkpush = "/mpesa/stkpush/v1/processrequest";

  /// Stk query endpoint
  static const String stkquery = "/mpesa/stkpushquery/v1/query";
}


lib/src/mpesa_base.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../utils/routes.dart';

/// Initializes a new instance of [Mpesa]
/// Requires 4 parameters:
///
/// 1. `clientKey` - This is your consumer key
///
/// 2. `clientSecret` - This is your consumer secret
///
/// 3. `environment` - This is the environment our app is running on. It can either be `sandbox` or `production`.
///
/// 4. `initiatorPassword` - You can get this from Your Portal(production) or from test credentials(Sandbox). It will be the Security Credential (Shortcode 1)
class Mpesa {
  final String _baseUrl;

  /// Your consumer key
  final String clientKey;

  /// Your consumer secret
  final String clientSecret;

  /// Environment the app is running on. It can either be `sandbox` or `production`
  final String environment;
  final String _accessToken;
  final String _passKey;

  /// Initializes a new instance of [Mpesa]
  /// Requires 4 parameters:
  ///
  /// 1. `clientKey` - This is your consumer key
  ///
  /// 2. `clientSecret` - This is your consumer secret
  ///
  /// 3. `environment` - This is the environment our app is running on. It can either be `sandbox` or `production`.
  ///
  /// 4. `initiatorPassword` - You can get this from Your Portal(production) or from test credentials(Sandbox). It will be the Security Credential (Shortcode 1)
  Mpesa({
    required this.clientKey,
    required this.clientSecret,
    required this.environment,
    required String passKey,
  })  : assert(environment == "production" || environment == "sandbox"),
        _passKey = passKey,
        _baseUrl =
            environment == "production" ? Routes.production : Routes.sandbox,
        _accessToken = base64Url.encode("$clientKey:$clientSecret".codeUnits) {
    // _generateSecurityCredential(password: "", certificatepath: null);
  }

  Future<String> _authenticate() async {
    try {
      final http.Response response = await http.get(
        Uri.parse("$_baseUrl${Routes.oauth}"),
        headers: {
          "Authorization": "Basic $_accessToken",
        },
      );

      final data = json.decode(response.body);

      return data["access_token"] as String;
    } catch (e) {
      rethrow;
    }
  }

  /// Triggers a lipa na mpesa stk push and presents user with dialog to input mpesa pin. The method will complete regardless of whether the transaction was successful or not. Results of the transaction are sent to the [callbackUrl] provided. Ensure that the [callbackUrl] provided is publicly accessible. You can use ngrok,localtunnel or serveo for local development.
  ///
  /// `businessShortCode` can be gotten from https://developer.safaricom.co.ke/APIs/MpesaExpressSimulate under the Lipa Na Mpesa Online simulator under PartyB. You can ignore this if you have already set it in [globalBusinessShortCode]. Please note that this is for the sandbox environment. Use your registered business short code for production.
  ///
  /// `phoneNumber` is the phone number(MSISDN) to be charged. It must be a registered mpesa number(MSISDN) and should contain the international dialing code i.e `254`. For example `254712345678`
  ///
  /// `amount` is the amount to be charged. During development/sandbox all money transfered is refunded by safaricom within 24 hours. Please note that this is only applicable if you're using the [businessShortCode] provided by Safaricom and not a real one.
  ///
  /// `callbackUrl` is the url to which Mpesa responses are sent upon success or failure of a transaction. Should be able to receive post requests.
  ///
  /// `accountReference` used with Mpesa paybills as account number,
  ///
  ///  Please see https://developer.safaricom.co.ke/Documentation for more info.
  Future<MpesaResponse> lipaNaMpesa({
    required String phoneNumber,
    required double amount,
    required String callbackUrl,
    String? businessShortCode,
    String transactionType = "CustomerPayBillOnline",
    String accountReference = "account",
    String transactionDescription = "Lipa Na Mpesa Online",
  }) async {
    if (amount < 1.0) {
      throw "Amount should be at least Ksh 1";
    }

    final _now = DateTime.now();
    final _paybill = businessShortCode;

    if (_paybill == null) {
      throw "Paybill must be set either via businessShortCode or via globalBusinessShortCode";
    }

    final _timestamp =
        "${_now.year}${_now.month.toString().padLeft(2, '0')}${_now.day.toString().padLeft(2, '0')}${_now.hour.toString().padLeft(2, '0')}${_now.minute.toString().padLeft(2, '0')}${_now.second.toString().padLeft(2, '0')}";

    final String _rawPassword = _paybill + _passKey + _timestamp;

    final List<int> _passwordBytes = utf8.encode(_rawPassword);

    final String _password = base64.encode(_passwordBytes);

    final String token = await _authenticate();

    final String _body = json.encode({
      'BusinessShortCode': _paybill,
      'Password': _password,
      'Timestamp': _timestamp,
      'TransactionType': transactionType,
      'Amount': amount,
      'PartyA': phoneNumber,
      'PartyB': _paybill,
      'PhoneNumber': phoneNumber,
      'CallBackURL': callbackUrl.toString(),
      'AccountReference': accountReference,
      'TransactionDesc': transactionDescription
    });

    final http.Response response = await http.post(
      Uri.parse("$_baseUrl${Routes.stkpush}"),
      body: _body,
      headers: {
        'Authorization': 'Bearer $token',
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode != 200) {
      throw json.decode(response.body) as Map<dynamic, dynamic>;
    }

    return MpesaResponse.fromMap(
      json.decode(response.body) as Map<String, dynamic>,
    );
  }
}

/// Response return by Mpesa request. Please Note that this does not indicate the success or failure of the payment. That will be sent to the callback url you provided. This response only indicates whether the payment request was accepted for processing or not.
class MpesaResponse {
  /// The merchant request id
  // ignore: non_constant_identifier_names
  String MerchantRequestID;

  /// Unique Checkout request Id. This will be sent along with the success or failure data to the callback url you provide. You can use it to identify a particular payment request.
  // ignore: non_constant_identifier_names
  String CheckoutRequestID;

  /// M-Pesa Result and Response Codes
  ///
  /// `0` - Success
  ///
  /// `1` - Insufficient Funds
  ///
  /// `2` - Less Than Minimum Transaction Value
  ///
  /// `3` - More Than Maximum Transaction Value
  ///
  /// `4` - Would Exceed Daily Transfer Limit
  ///
  /// `5` - Would Exceed Minimum Balance
  ///
  /// `6` - Unresolved Primary Party
  ///
  /// `7` - Unresolved Receiver Party
  ///
  /// `8` - Would Exceed Maxiumum Balance
  ///
  /// `11` - Debit Account Invalid
  ///
  /// `12` - Credit Account Invalid
  ///
  /// `13` - Unresolved Debit Account
  ///
  /// `14` - Unresolved Credit Account
  ///
  /// `15` - Duplicate Detected
  ///
  /// `17` - Internal Failure
  ///
  /// `20` - Unresolved Initiator
  ///
  /// `26` - Traffic blocking condition in place
  ///
  // ignore: non_constant_identifier_names
  String ResponseCode;

  /// Description of the response gotten
  // ignore: non_constant_identifier_names
  String ResponseDescription;

  /// Message that is user friendly and can be shown to the customer.
  /// Mostly same as [ResponseDescription]
  // ignore: non_constant_identifier_names
  String CustomerMessage;

  /// Response return by Mpesa request. Please Note that this does not indicate the success or failure of the payment. That will be sent to the callback url you provided. This response only indicates whether the payment request was accepted for processing or not.
  MpesaResponse({
    // ignore: non_constant_identifier_names
    required this.MerchantRequestID,
    // ignore: non_constant_identifier_names
    required this.CheckoutRequestID,
    // ignore: non_constant_identifier_names
    required this.ResponseCode,
    // ignore: non_constant_identifier_names
    required this.ResponseDescription,
    // ignore: non_constant_identifier_names
    required this.CustomerMessage,
  });

  /// Create a new instance of [MpesaResponse] by copying values from another instance
  MpesaResponse copyWith({
    // ignore: non_constant_identifier_names
    String? MerchantRequestID,
    // ignore: non_constant_identifier_names
    String? CheckoutRequestID,
    // ignore: non_constant_identifier_names
    String? ResponseCode,
    // ignore: non_constant_identifier_names
    String? ResponseDescription,
    // ignore: non_constant_identifier_names
    String? CustomerMessage,
  }) {
    return MpesaResponse(
      MerchantRequestID: MerchantRequestID ?? this.MerchantRequestID,
      CheckoutRequestID: CheckoutRequestID ?? this.CheckoutRequestID,
      ResponseCode: ResponseCode ?? this.ResponseCode,
      ResponseDescription: ResponseDescription ?? this.ResponseDescription,
      CustomerMessage: CustomerMessage ?? this.CustomerMessage,
    );
  }

  /// [MpesaResponse] as a [Map]
  Map<String, dynamic> toMap() {
    return {
      'MerchantRequestID': MerchantRequestID,
      'CheckoutRequestID': CheckoutRequestID,
      'ResponseCode': ResponseCode,
      'ResponseDescription': ResponseDescription,
      'CustomerMessage': CustomerMessage,
    };
  }

  /// Create an instance of [MpesaResponse] from a [Map]
  factory MpesaResponse.fromMap(Map<String, dynamic> map) {
    return MpesaResponse(
      MerchantRequestID: map['MerchantRequestID'] as String,
      CheckoutRequestID: map['CheckoutRequestID'] as String,
      ResponseCode: map['ResponseCode'] as String,
      ResponseDescription: map['ResponseDescription'] as String,
      CustomerMessage: map['CustomerMessage'] as String,
    );
  }

  @override
  String toString() {
    return 'MpesaResponse(MerchantRequestID: $MerchantRequestID, CheckoutRequestID: $CheckoutRequestID, ResponseCode: $ResponseCode, ResponseDescription: $ResponseDescription, CustomerMessage: $CustomerMessage)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is MpesaResponse &&
        other.MerchantRequestID == MerchantRequestID &&
        other.CheckoutRequestID == CheckoutRequestID &&
        other.ResponseCode == ResponseCode &&
        other.ResponseDescription == ResponseDescription &&
        other.CustomerMessage == CustomerMessage;
  }

  @override
  int get hashCode {
    return MerchantRequestID.hashCode ^
        CheckoutRequestID.hashCode ^
        ResponseCode.hashCode ^
        ResponseDescription.hashCode ^
        CustomerMessage.hashCode;
  }
}
